### PR TITLE
cluster: correct ipv6 url for strict dns cluster and logical dns cluster

### LINKED ISSUE #18606: literal IPv6 addresses are not parsed correctly in lb_endpoints with STRICT_DNS
I want to have cluster lb_endpoints that may include both hostnames and literal IP addresses at the same time. When the cluster's type is `STRICT_DNS`, the configuration is refused with this error keeps iterating in the logs
```
[2021-10-13 13:53:47.264][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.cluster.v3.Cluster rejected: Error adding/updating cluster(s) default-c02-tcp-443-cluster: malformed url: tcp://fdee:6207::1000:0:0:a3bd:22607
```

I guess this is because the square brackets are not set around the IPv6 literal (i.e. it should be `tcp://[fdee:6207::1000:0:0:a3bd]:22607`)? I tried having IPv4 literals and they work perfectly fine. So this is an IPv6-specific issue.

Also I tried it with `LOGICAL_DNS` but I got the error
```
[2021-10-13 13:33:38.163][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.cluster.v3.Cluster rejected: Error adding/updating cluster(s) default-c02-tcp-443-cluster: LOGICAL_DNS clusters must have a single locality_lb_endpoint and a single lb_endpoint
```
But that was expected since I have more than one endpoint.

Also I am using Envoy 1.19 if that may help.

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: correct ipv6 url for strict dns cluster and logical dns cluster
Additional Description:
Strict DNS cluster and logical DNS cluster building a wrong URL with ipv6 address as `tcp://::1:5000`, it leads an exception raised from the cluster config parsing. The correct URL should be `tcp://[::1]:5000`.

Risk Level: low
Testing: unittest added
Docs Changes: n/a
Release Notes: n/a
Fixes #18606

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

### CODE REVIEW

File: source/common/network/utility.cc [hunk 1]
```diff
@@ -66,13 +66,30 @@ std::string hostFromUrl(const std::string& url, absl::string_view scheme,
...
   }
 
-  const size_t colon_index = url.find(':', scheme.size());
+  const size_t bracket_begin_index = url.find('[', scheme.size());
 
-  if (colon_index == std::string::npos) {
-    throw EnvoyException(absl::StrCat("malformed url: ", url));
-  }
+  if (bracket_begin_index == std::string::npos) {
+    const size_t colon_index = url.find(':', scheme.size());
```
Comment (snowp) [inline source/common/network/utility.cc:72]: Could you include some comments here that explains the logic a bit? It's a bit hard to follow just from the code

File: source/common/network/utility.cc [hunk 2]
```diff
@@ -66,13 +66,30 @@ std::string hostFromUrl(const std::string& url, absl::string_view scheme,
...
+
+    if ((colon_index == std::string::npos) || (colon_index == scheme.size())) {
```
Comment (snowp) [inline source/common/network/utility.cc:74]: Why does colon_index == scheme.size() indicate a malformed URL? Comment would be good

File: source/common/network/utility.h [hunk 3]
```diff
@@ -121,6 +121,14 @@ class Utility {
    */
   static bool urlIsUnixScheme(absl::string_view url);
 
+  /**
+   * Build a TCP URL
```
Comment (snowp) [inline source/common/network/utility.h:125]: Can you include some documentation here that this function handles formatting IPv6?

File: source/common/network/utility.h [hunk 4]
```diff
@@ -121,6 +121,14 @@ class Utility {
...
+   * @param ip_address as ip address include in URL
+   * @param port to included in URL.
+   * @return std::string the tcp URL
+   */
+  static std::string getTcpUrl(const std::string& ip_address, uint16_t port);
```
Comment (snowp) [inline source/common/network/utility.h:130]: maybe `formatTcpUrl`?

**Review by snowp — CHANGES_REQUESTED** [review]
A few style comments, I think this makes sense

Would be nice if someone that's spent more time working with IPv6 could look over the parsing code, maybe @RyanTheOptimist ?

**Comment by soulxu** [comment]
> Would be nice if someone that's spent more time working with IPv6 could look over the parsing code, maybe @RyanTheOptimist ?

Would you like to elaborate on what parsing code you pointed to? I may help on it also

File: source/common/network/utility.cc (hunk 1 above)
Comment (soulxu) [reply source/common/network/utility.cc:72]: got it, let me add the comment, thanks for the review!

File: source/common/network/utility.h (hunk 4 above)
Comment (soulxu) [reply source/common/network/utility.h:130]: yea, that is better name. thanks!

File: source/common/network/utility.cc [hunk 5]
```diff
@@ -66,13 +66,37 @@ std::string hostFromUrl(const std::string& url, absl::string_view scheme,
...
   }
 
-  const size_t colon_index = url.find(':', scheme.size());
+  const size_t bracket_begin_index = url.find('[', scheme.size());
 
-  if (colon_index == std::string::npos) {
-    throw EnvoyException(absl::StrCat("malformed url: ", url));
-  }
+  // If there is no '[', it means it is ipv4 address, otherwise it is ipv6 address.
+  if (bracket_begin_index == std::string::npos) {
```
Comment (RyanTheOptimist) [inline source/common/network/utility.cc:72]: Hm. I'm slightly apprehensive about doing all this manual URL parsing since, as this PR points out, it's easy to get wrong. Would it be possible to instead use the google URL library which is already an envoy dependency?

Using this, I think it should be roughly as simple as:
```
return GURL(url).host();
```
(Modulo error checking, etc). For example, here's the GURL code which handles IP address canonicalization:

https://quiche.googlesource.com/googleurl/+/refs/heads/master/url/url_canon_ip.cc

Comment (soulxu) [reply source/common/network/utility.cc:72]: Yea, let me take a look at. Thanks for the review

Comment (soulxu) [reply source/common/network/utility.cc:72]: @RyanTheOptimist I did a test, seems `GURL` doesn't support the scheme of `tcp` and `udp` .

https://quiche.googlesource.com/googleurl/+/refs/heads/master/url/url_constants.cc

Comment (dio) [reply source/common/network/utility.cc:72]: Should we add those schemes to that library? cc. @htuch @yanavlasov

Comment (htuch) [reply source/common/network/utility.cc:72]: Yes, that seems the cleanest thing, or at least make it easy to add additional schemes.

Comment (soulxu) [reply source/common/network/utility.cc:72]: Do you mean contribute that back to chromium commuinity?

Comment (yanavlasov) [reply source/common/network/utility.cc:72]: We unfortunately can not use GURL directly because of ICU dependency. But you should be able to use this class https://github.com/envoyproxy/envoy/blob/05bdb1d6850da7b00f1ee3b0f4d8bc2f3131f7c9/source/common/http/utility.h#L144 that should deal with the IPv6 hostnames correctly.

Comment (htuch) [reply source/common/network/utility.cc:72]: Oh right, we have an independent Url class still, that's much better, we can hide the details later as needed if we switch to GURL there.

Comment (soulxu) [reply source/common/network/utility.cc:72]: thanks for the feedback, let me update

**Comment by soulxu** [comment]
I just need a time to update it
